clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```
@@ -653,6 +694,85 @@ public void onFailure(RuntimeException e) {
+                        if (!currentLeader.equals(cachedLeaderAndEpochs.get(respTopicPartition))) {
+                            return;
+                        }
+
+
+                        if (subscriptions.awaitingValidation(respTopicPartition)) {
+                            SubscriptionState.FetchPosition currentPosition = subscriptions.position(respTopicPartition);
+                            if (respEndOffset.endOffset() < currentPosition.offset) {
+                                if (subscriptions.hasDefaultOffsetResetPolicy()) {
+                                    SubscriptionState.FetchPosition newPosition = new SubscriptionState.FetchPosition(
+                                            respEndOffset.endOffset(), Optional.of(respEndOffset.leaderEpoch()), currentLeader);
+                                    log.info("Truncation detected for partition {}, resetting offset to {}", respTopicPartition, newPosition);
+                                    subscriptions.position(respTopicPartition, newPosition);
+                                } else {
+                                    log.warn("Truncation detected for partition {}, but no reset policy is set", respTopicPartition);
+                                    truncationWithoutResetPolicy.put(respTopicPartition, new OffsetAndMetadata(
+                                            respEndOffset.endOffset(), Optional.of(respEndOffset.leaderEpoch()), null));
+                                }
+                            } else {
+                                // Offset is fine, clear the validation state
+                                subscriptions.validate(respTopicPartition);
+                            }
+                        }
+                    });
+
+                    if (!truncationWithoutResetPolicy.isEmpty()) {
+                        throw new LogTruncationException(truncationWithoutResetPolicy);
```

If the user ignores this exception and calls `poll()`, will we repeat the validation? That may be reasonable, just making sure I understand.

---

At this point, the subscription state is still in AWAITING_VALIDATION, so yes the next `poll()` call should attempt to revalidate asynchronously and these partitions will be excluded for the time being.